File: leetcode/Searcha2DMatrix/searchMatrix.java
public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        int height = matrix.length;
        if (height == 0) return false;
        int width = matrix[0].length, l = 0, r = height - 1, m;
        while (l <= r) {
            m = l + (r - l >> 1);
            if (matrix[m][0] == target) return true;
            if (matrix[m][0] > target) r = m - 1;
            else l = m + 1;
        }
        if (r < 0) return false;
        int row = r;
        l = 0;
        r = width - 1;
        while (l <= r) {
            m = l + (r - l >> 1);
            if (matrix[row][m] == target) return true;
            if (matrix[row][m] > target) r = m - 1;
            else l = m + 1;
        }
        return false;
    }
}

File: leetcode/Searcha2DMatrix/searchMatrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = m ? matrix[0].size() : 0, i = 0, j = m * n - 1;
        while (i <= j) {
            int mid = i + (j - i) / 2, row = mid / n, col = mid % n;
            if (matrix[row][col] == target) return true;
            else if (matrix[row][col] < target) i = mid + 1;
            else j = mid - 1;
        }
        return false;
    }
};
